x-pack/plugins/security_solution/cypress/e2e/detection_rules/persistent_rules_table_state.cy.ts

```diff
@@ -0,0 +1,489 @@
+  createRule('8', 'rule 7', ['tag-b'], true);
+}
+
+function visitRulesTableWithState(urlTableState: Record<string, unknown>): void {
+  visit(SECURITY_DETECTIONS_RULES_MANAGEMENT_URL, { qs: { rulesTable: encode(urlTableState) } });
+}
+
+function setStorageState(storageTableState: Record<string, unknown>): void {
+  cy.window().then((win) => {
+    win.sessionStorage.setItem('securitySolution.rulesTable', JSON.stringify(storageTableState));
+  });
+}
+
+function changeRulesTableState(): void {
+  filterBySearchTerm('rule');
+  filterByTags(['tag-b']);
+  filterByCustomRules();
+  filterByDisabledRules();
+  sortByTableColumn('Rule', 'asc');
+  setRowsPerPageTo(5);
+}
+
+function expectRulesTableState(): void {
+  expectFilterSearchTerm('rule');
+  expectFilterByTags(['tag-b']);
+  expectFilterByCustomRules();
+  expectFilterByDisabledRules();
+  expectTableSorting('Rule', 'asc');
+  expectRowsPerPage(5);
+}
+
+function expectDefaultRulesTableState(): void {
+  expectFilterSearchTerm('');
+  expectNoFilterByTags();
+  expectNoFilterByElasticOrCustomRules();
+  expectNoFilterByEnabledOrDisabledRules();
+  expectTableSorting('Enabled', 'desc');
+  expectRowsPerPage(20);
+  expectTablePage(1);
+}
+
+function expectManagementTableRules(ruleNames: string[]): void {
+  expectNumberOfRules(RULES_MANAGEMENT_TABLE, ruleNames.length);
+
+  for (const ruleName of ruleNames) {
+    expectToContainRule(RULES_MANAGEMENT_TABLE, ruleName);
+  }
+}
+
+function expectMonitoringTableRules(ruleNames: string[]): void {
+  expectNumberOfRules(RULES_MONITORING_TABLE, ruleNames.length);
+
+  for (const ruleName of ruleNames) {
+    expectToContainRule(RULES_MONITORING_TABLE, ruleName);
+  }
+}
+
+describe('Persistent rules table state', () => {
+  before(() => {
+    cleanKibana();
+    createTestRules();
+    login();
+  });
+
+  beforeEach(() => {
+    resetRulesTableState();
+  });
+
+  describe('while on a happy path', () => {
+    it('activates management tab by default', () => {
+      visit(SECURITY_DETECTIONS_RULES_URL);
+
+      expectRulesManagementTab();
+    });
+
+    it('prefers url state over storage state', () => {
+      setStorageState({
+        searchTerm: 'test',
+        tags: ['tag-c'],
+        source: 'prebuilt',
+        enabled: true,
+        field: 'severity',
+        order: 'desc',
+        perPage: 10,
+      });
+
+      visitRulesTableWithState({
+        searchTerm: 'rule',
+        tags: ['tag-b'],
+        source: 'custom',
+        enabled: false,
+        field: 'name',
+        order: 'asc',
+        perPage: 5,
+        page: 2,
+      });
+
+      expectRulesManagementTab();
+      expectRulesTableState();
+      expectTablePage(2);
+      expectManagementTableRules(['rule 6']);
+    });
+
+    describe('and on the rules management tab', () => {
+      beforeEach(() => {
+        visit(SECURITY_DETECTIONS_RULES_MANAGEMENT_URL);
+
+        changeRulesTableState();
+        goToTablePage(2);
```

Comment:
Let's transfer these two preconditions into the test case bodies to improve the readability of tests. When initial conditions and checks are separated, reading and understanding the tests can be challenging. People typically read text from top to bottom, and the test structure should ideally adhere to the "Given-When-Then" pattern:

```gherkin
Given something
When action
Then result
```

By placing the initial conditions in the `beforeEach` section, developers are compelled to switch back and forth between it and the test body to comprehend what is being verified. The `beforeEach` section is suitable for clearing the state or setting up some relatively insignificant initial conditions.

Reply:
done